File: eater-backend/src/api/services/meal.log.service.js
const MealLog = require("../../models/meal_log");

/**
 * Meal Log Service
 * Handles all business logic for meal logging operations
 */
class MealLogService {
  /**
   * Save a new meal log entry
   * @param {String} userId - User ID from JWT token
   * @param {Object} mealLogData - Meal log data object
   * @returns {Promise<Object>} Saved meal log with id field
   * @throws {Error} If validation fails or save fails
   */
  async saveMealLog(userId, mealLogData) {
    const {
      mealName,
      mealType,
      calories,
      protein,
      carbs,
      fats,
      quantity,
      unit,
      notes,
      loggedAt,
      mealPlanId,
      mealPlanItemId,
      imageUrl,
    } = mealLogData;

    // Validate required fields
    if (!mealName || !mealType || calories === undefined || calories === null || 
        quantity === undefined || quantity === null || !unit || !loggedAt) {
      throw new Error(
        "Missing required fields: mealName, mealType, calories, quantity, unit, loggedAt"
      );
    }

    const mealLog = new MealLog({
      userId,
      mealName,
      mealType,
      calories: Number(calories),
      protein: Number(protein) || 0,
      carbs: Number(carbs) || 0,
      fats: Number(fats) || 0,
      quantity: Number(quantity),
      unit,
      notes: notes || null,
      imageUrl: imageUrl || null,
      loggedAt: new Date(loggedAt),
      mealPlanId: mealPlanId || null,
      mealPlanItemId: mealPlanItemId || null,
    });

    await mealLog.save();

    console.log(
      `[MealLogService.saveMealLog] Saved - ID: ${mealLog._id}, userId: ${userId}, mealType: ${mealType}`
    );

    return {
      id: mealLog._id.toString(),
      ...mealLog.toObject(),
    };
  }

  /**
   * Get all meal logs for a user with pagination
   * @param {String} userId - User ID from JWT token
   * @param {Object} query - Query object with page and limit
   * @returns {Promise<Object>} Array of meal logs with pagination info
   */
  async getAllMealLogs(userId, query) {
    const page = parseInt(query.page) || 1;
    const limit = parseInt(query.limit) || 50;
    const skip = (page - 1) * limit;

    const mealLogs = await MealLog.find({ userId })
      .sort({ loggedAt: -1 })
      .skip(skip)
      .limit(limit);

    const total = await MealLog.countDocuments({ userId });

    return {
      meals: mealLogs.map((log) => ({
        id: log._id.toString(),
        ...log.toObject(),
      })),
      pagination: {
        page,
        limit,
        total,
        totalPages: Math.ceil(total / limit),
      },
    };
  }

  /**
   * Get meal logs for a specific date
   * @param {String} userId - User ID from JWT token
   * @param {String} date - Date string in YYYY-MM-DD format
   * @returns {Promise<Object>} Meal logs and daily totals for the date
   */
  async getMealLogsForDate(userId, date) {
    if (!date) {
      throw new Error("Date parameter is required (format: YYYY-MM-DD)");
    }

    const startDate = new Date(date);
    startDate.setHours(0, 0, 0, 0);

    const endDate = new Date(date);
    endDate.setHours(23, 59, 59, 999);

    const mealLogs = await MealLog.find({
      userId,
      loggedAt: {
        $gte: startDate,
        $lte: endDate,
      },
    }).sort({ loggedAt: 1 });

    // Calculate daily totals
    const totals = {
      calories: 0,
      protein: 0,
      carbs: 0,
      fats: 0,
    };

    mealLogs.forEach((log) => {
      totals.calories += log.calories;
      totals.protein += log.protein;
      totals.carbs += log.carbs;
      totals.fats += log.fats;
    });

    return {
      meals: mealLogs.map((log) => ({
        id: log._id.toString(),
        ...log.toObject(),
      })),
      totals,
      date,
    };
  }

  /**
   * Get meal logs for a date range
   * @param {String} userId - User ID from JWT token
   * @param {String} startDate - Start date string in YYYY-MM-DD format
   * @param {String} endDate - End date string in YYYY-MM-DD format
   * @returns {Promise<Object>} Meal logs for the date range
   */
  async getMealLogsForDateRange(userId, startDate, endDate) {
    if (!startDate || !endDate) {
      throw new Error("Both startDate and endDate are required (format: YYYY-MM-DD)");
    }

    const start = new Date(startDate);
    start.setHours(0, 0, 0, 0);

    const end = new Date(endDate);
    end.setHours(23, 59, 59, 999);

    const mealLogs = await MealLog.find({
      userId,
      loggedAt: {
        $gte: start,
        $lte: end,
      },
    }).sort({ loggedAt: -1 });

    return {
      meals: mealLogs.map((log) => ({
        id: log._id.toString(),
        ...log.toObject(),
      })),
      count: mealLogs.length,
      dateRange: {
        startDate,
        endDate,
      },
    };
  }

  /**
   * Update a meal log entry
   * @param {String} mealLogId - Meal log ID to update
   * @param {String} userId - User ID from JWT token (for ownership verification)
   * @param {Object} updateData - Data to update
   * @returns {Promise<Object>} Updated meal log
   * @throws {Error} If not found or not authorized
   */
  async updateMealLog(mealLogId, userId, updateData) {
    if (!mealLogId) {
      throw new Error("Meal log ID is required");
    }

    const {
      mealName,
      mealType,
      calories,
      protein,
      carbs,
      fats,
      quantity,
      unit,
      notes,
      imageUrl,
    } = updateData;

    // Validate required fields
    if (!mealName || !mealType || calories === undefined || calories === null || 
        quantity === undefined || quantity === null || !unit) {
      throw new Error(
        "Missing required fields: mealName, mealType, calories, quantity, unit"
      );
    }

    // Find the meal log by ID
    let mealLog;
    try {
      mealLog = await MealLog.findById(mealLogId);
    } catch (findErr) {
      throw new Error(`Invalid meal log ID format: ${findErr.message}`);
    }

    if (!mealLog) {
      throw new Error("Meal log not found");
    }

    // Verify ownership
    if (mealLog.userId && mealLog.userId.toString() !== userId) {
      throw new Error("Not authorized to update this meal log");
    }

    // Update fields
    mealLog.mealName = mealName;
    mealLog.mealType = mealType;
    mealLog.calories = Number(calories);
    mealLog.protein = Number(protein) || 0;
    mealLog.carbs = Number(carbs) || 0;
    mealLog.fats = Number(fats) || 0;
    mealLog.quantity = Number(quantity);
    mealLog.unit = unit;
    mealLog.notes = notes || null;
    mealLog.imageUrl = imageUrl || null;

    // Set userId if not already set
    if (!mealLog.userId) {
      mealLog.userId = userId;
    }

    await mealLog.save();

    console.log(`[MealLogService.updateMealLog] Updated - ID: ${mealLogId}`);

    return {
      id: mealLog._id.toString(),
      ...mealLog.toObject(),
    };
  }

  /**
   * Delete a meal log entry
   * @param {String} mealLogId - Meal log ID to delete
   * @param {String} userId - User ID from JWT token (for ownership verification)
   * @returns {Promise<void>}
   * @throws {Error} If not found or not authorized
   */
  async deleteMealLog(mealLogId, userId) {
    if (!mealLogId) {
      throw new Error("Meal log ID is required");
    }

    const mealLog = await MealLog.findByIdAndDelete(mealLogId);

    if (!mealLog) {
      throw new Error("Meal log not found");
    }

    // Verify ownership
    if (mealLog.userId && mealLog.userId.toString() !== userId) {
      throw new Error("Not authorized to delete this meal log");
    }

    console.log(`[MealLogService.deleteMealLog] Deleted - ID: ${mealLogId}`);
  }
}

module.exports = new MealLogService();

File: mobile/lib/src/features/nutrition/data/services/meal_log_service.dart
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';

import '../../domain/entities/meal_log.dart';

class MealLogService {
  final Dio dio;

  MealLogService({required this.dio});

  static const String baseUrl = '/api/meal-logs';

  /// Save a meal log to the backend
  /// Returns the saved meal log with ID from the server
  Future<MealLog> saveMealLog(MealLog mealLog) async {
    try {
      debugPrint(
          '[saveMealLog] Creating meal with loggedAt: ${mealLog.loggedAt.toString()}');

      // Format date string WITHOUT timezone conversion
      // Extract year/month/day from local DateTime, don't use toIso8601String() which converts to UTC
      final year = mealLog.loggedAt.year;
      final month = mealLog.loggedAt.month.toString().padLeft(2, '0');
      final day = mealLog.loggedAt.day.toString().padLeft(2, '0');
      final loggedAtString = '$year-$month-${day}T00:00:00.000Z';
      debugPrint(
          '[saveMealLog] Formatted loggedAt string (no timezone conversion): $loggedAtString');

      final response = await dio.post(
        baseUrl,
        data: {
          'mealName': mealLog.mealName,
          'mealType': mealLog.mealType,
          'calories': mealLog.calories,
          'protein': mealLog.protein,
          'carbs': mealLog.carbs,
          'fats': mealLog.fats,
          'quantity': mealLog.quantity,
          'unit': mealLog.unit,
          'notes': mealLog.notes,
          'imageUrl': mealLog.imageUrl,
          'loggedAt': loggedAtString,
        },
      );

      if (response.statusCode == 201 || response.statusCode == 200) {
        // Debug: Print raw response
        debugPrint('[saveMealLog] Response status: ${response.statusCode}');
        debugPrint('[saveMealLog] Full response: ${response.data}');

        // Return the saved meal log (server may return additional fields like ID)
        final data = response.data['data'] ?? response.data;
        debugPrint('[saveMealLog] Parsed data: $data');

        final savedMeal = MealLog.fromJson(data as Map<String, dynamic>);
        debugPrint(
            '[saveMealLog] Parsed MealLog - ID: ${savedMeal.id}, name: ${savedMeal.mealName}, loggedAt: ${savedMeal.loggedAt}');

        return savedMeal;
      } else {
        throw Exception('Failed to save meal log: ${response.statusCode}');
      }
    } on DioException catch (e) {
      debugPrint('[saveMealLog] DioException: ${e.message}');
      throw Exception('Error saving meal log: ${e.message}');
    } catch (e) {
      debugPrint('[saveMealLog] Exception: $e');
      throw Exception('Unexpected error saving meal log: $e');
    }
  }

  /// Update an existing meal log
  /// Returns the updated meal log
  Future<MealLog> updateMealLog(String mealLogId, MealLog mealLog) async {
    try {
      final response = await dio.put(
        '$baseUrl/$mealLogId',
        data: {
          'mealName': mealLog.mealName,
          'mealType': mealLog.mealType,
          'calories': mealLog.calories,
          'protein': mealLog.protein,
          'carbs': mealLog.carbs,
          'fats': mealLog.fats,
          'quantity': mealLog.quantity,
          'unit': mealLog.unit,
          'notes': mealLog.notes,
          'imageUrl': mealLog.imageUrl,
        },
      );

      if (response.statusCode == 200) {
        // Return the updated meal log
        final data = response.data['data'] ?? response.data;
        return MealLog.fromJson(data as Map<String, dynamic>);
      } else {
        throw Exception('Failed to update meal log: ${response.statusCode}');
      }
    } on DioException catch (e) {
      throw Exception('Error updating meal log: ${e.message}');
    } catch (e) {
      throw Exception('Unexpected error updating meal log: $e');
    }
  }

  /// Get meal logs for a specific date
  Future<List<MealLog>> getMealLogsForDate(DateTime date) async {
    try {
      final dateStr =
          '${date.year}-${date.month.toString().padLeft(2, '0')}-${date.day.toString().padLeft(2, '0')}';
      final response = await dio.get('$baseUrl/date/$dateStr');

      if (response.statusCode == 200) {
        final data = response.data['data'] as Map<String, dynamic>;
        final mealsData = data['meals'] as List<dynamic>? ?? [];
        return mealsData
            .map((m) => MealLog.fromJson(m as Map<String, dynamic>))
            .toList();
      } else {
        throw Exception('Failed to fetch meal logs: ${response.statusCode}');
      }
    } on DioException catch (e) {
      throw Exception('Error fetching meal logs: ${e.message}');
    } catch (e) {
      throw Exception('Unexpected error fetching meal logs: $e');
    }
  }

  /// Get all meal logs with pagination
  Future<List<MealLog>> getAllMealLogs({int page = 1, int limit = 50}) async {
    try {
      final response = await dio.get(
        baseUrl,
        queryParameters: {
          'page': page,
          'limit': limit,
        },
      );

      if (response.statusCode == 200) {
        final data = response.data['data'] as Map<String, dynamic>;
        final mealsData = data['meals'] as List<dynamic>? ?? [];
        return mealsData
            .map((m) => MealLog.fromJson(m as Map<String, dynamic>))
            .toList();
      } else {
        throw Exception('Failed to fetch meal logs: ${response.statusCode}');
      }
    } on DioException catch (e) {
      throw Exception('Error fetching meal logs: ${e.message}');
    } catch (e) {
      throw Exception('Unexpected error fetching meal logs: $e');
    }
  }

  /// Delete a meal log
  Future<bool> deleteMealLog(String mealLogId) async {
    try {
      final response = await dio.delete('$baseUrl/$mealLogId');

      if (response.statusCode == 200) {
        return true;
      } else {
        throw Exception('Failed to delete meal log: ${response.statusCode}');
      }
    } on DioException catch (e) {
      throw Exception('Error deleting meal log: ${e.message}');
    } catch (e) {
      throw Exception('Unexpected error deleting meal log: $e');
    }
  }
}
